### PR TITLE
Update CHANGES for 8.6beta1.

### PR DESCRIPTION
**Pull request by maximedenes**
Please let me know quickly if I forgot something to put here.

### CODE REVIEW

File: CHANGES [hunk 1]
```diff
@@ -110,6 +101,11 @@ Notations
 
 Tools
 
+- New warnings system for coqtop/coqc. Can be configured with the flag "-w" or
+  with the Vernacular command "Set Warnings".
+- Warnings are now printed by coqc.
+- In asynchronous mode, Coq is now capable of recovering from errors and
+  continue processing the document.
 - coqc accepts a -o option to specify the output file name
```
Comment (herbelin) [inline CHANGES]: We used an "Internal infrastructure" category for this kind of change.

File: CHANGES [hunk 2]
```diff
@@ -1,22 +1,10 @@
...
-  (i.e. non-minimizable).
-- #4592, #4932: notations sharing recursive patterns or sharing
-  binders made more robust.
-- #4780: Induction with universe polymorphism on was creating ill-typed terms.
-- #3070: fixing "subst" in the presence of a chain of dependencies.
-- When used as an argument of an ltac function, "auto" without "with"
-  nor "using" clause now correctly uses only the core hint database by
-  default.
+- A new, state-of-the-art universe constraint checker that
+  can outperform the previous version by an order of magnitude.
```
Comment (herbelin) [inline CHANGES]: Why "can"?

Being more explicit that "outperform" is about the efficiency/complexity of the algorithm?

File: CHANGES [hunk 3]
```diff
@@ -60,7 +48,7 @@ Tactics
   given a free identifier, it is not bound in subsequent tactics anymore.
   In order to introduce a binding, use e.g. the "fresh" primitive instead
   (potential source of incompatibilities).
-- New tactics is_ind, is_const, is_proj, is_constructor for use in Ltac (DOC TODO).
+- New tactics is_ind, is_const, is_proj, is_constructor for use in Ltac.
 - New goal selectors.  Sets of goals can be selected by select by listing
   integers ranges. Example: "1,4-7,24: tac" focuses "tac" on goals 1,4,5,6,7,24.
```
Comment (herbelin) [inline CHANGES]: "by select" to remove?

File: CHANGES [hunk 4]
```diff
@@ -78,6 +66,9 @@ Tactics
 - New tactics "notypeclasses refine" and "simple notypeclasses refine" that
   disallow typeclass resolution when typechecking their argument, for use
   in typeclass hints.
+- Integration of LtacProf, a profiler for Ltac.
+- Reduction tactics now accept more fine-grained flags: iota is now a shorthand
+  for match, fix and cofix.
```
Comment (herbelin) [inline CHANGES]: Explaining that "match", "fix" and "cofix" are new?

File: CHANGES (hunk 1 above)
Comment (maximedenes) [inline CHANGES]: For what of changes you mean? These are changes that users can see, not really internal, right?

File: CHANGES (hunk 2 above)
Comment (maximedenes) [inline CHANGES]: Alright, I'll simplify the wording. This is taken from credits btw.

File: CHANGES (hunk 1 above)
Comment (herbelin) [inline CHANGES]: For what of changes you mean? For the warnings.

Not really internal, right? That's right that it is visible. So, maybe "tools" is indeed the best approximation if one considers categories already used in the CHANGES file (even if, to tell the truth, I hardly see "coqtop" and "coqc" as tools, in contrast to "coqdep", "coqdoc", "coq_makefile", ...)

Otherwise, maybe "general infrastructure", since the warning infrastructure is really independent in itself of "coqc", "coqtop".

File: CHANGES [hunk 5]
```diff
@@ -120,6 +115,123 @@ Tools
   Verbose Compat vernacular, since these warnings can now be silenced or
   turned into errors using "-w".
 
+Bugfixes
+
```
Comment (herbelin) [inline CHANGES]: Is the list exhaustive (at least w.r.t explicitly reported bugs)? Matthieu said there were more than 450 bugs fixed.

Maybe "Excerpt of bug fixes" or "Selected list of bug fixes", depending on your intention.

File: CHANGES (hunk 1 above)
Comment (maximedenes) [inline CHANGES]: Well for 8.3, you mentioned the beautifier, -R and compatibility options to coqc/coqtop in the "tools" category :)

File: CHANGES (hunk 5 above)
Comment (herbelin) [inline CHANGES]: > Maybe I should simply write a last item saying that there are other fixes not mentioned here.

Seems reasonable.

However, more generally, what is your intent with the list of bugs? I mean, the list of bug fixes (and actually also of enhancements) is a very valuable information, but a request to the bug-tracker may be as much informative while being at the same time more customizable.

For instance, [a search on bugs with the FIXED tag and resolution date after Sep 2015](https://coq.inria.fr/bugs/buglist.cgi?f1=resolution&list_id=234517&o1=changedafter&order=Bug%20Number&query_format=advanced&resolution=FIXED&v1=2015-9-01) gives 424 bugs which altogether seem to include the list that you gave (and which goes closer to the 450 bugs counted by Matthieu).

So maybe the list could be replaced by a link to some well-chosen request listing the bugs, or by a copy-paste of the result of this request?

Comment (mattam82) [inline CHANGES]: Good point, that could solve the problem entirely! What I searched for exactly was this:
https://coq.inria.fr/bugs/buglist.cgi?product=Coq&field0-0-0=resolution&type0-0-0=equals&value0-0-0=FIXED&field0-0-1=resolution&type0-0-1=equals&value0-0-1=INVALID&field0-0-2=resolution&type0-0-2=equals&value0-0-2=WONTFIX&field0-0-3=resolution&type0-0-3=equals&value0-0-3=DUPLICATE&field0-1-0=delta_ts&type0-1-0=lessthan&value0-1-0=2016-11-06&field0-2-0=delta_ts&type0-2-0=greaterthan&value0-2-0=2016-01-01

Comment (mattam82) [inline CHANGES]: A query that can be run without login: https://coq.inria.fr/bugs/buglist.cgi?bug_status=RESOLVED&bug_status=CLOSED&chfield=resolution&chfieldfrom=2016-01-01&chfieldto=Now&query_format=advanced&resolution=FIXED&resolution=INVALID&resolution=WONTFIX&resolution=DUPLICATE&resolution=WORKSFORME&resolution=MOVED

Comment (mattam82) [inline CHANGES]: This gives 459 bugs

Comment (maximedenes) [inline CHANGES]: But these are not necessarily bugs fixed in 8.6, are they?

Comment (maximedenes) [inline CHANGES]: @herbelin I don't know exactly what the intent of this list is. I did it a bit the same way as the ones for earlier versions. We should indeed discuss it at some point, but maybe we can do so after the beta. Note that the title of the bug is not always the description of the change. I think the list should list the changes, rather than the original symptom. But I'm not sure.

File: CHANGES (hunk 1 above)
Comment (maximedenes) [inline CHANGES]: Sound good to me, I'll do it and merge.

**Comment by maximedenes** [comment]
Improved, rebased and merged. Thanks for the feedback.

File: CHANGES (hunk 5 above)
Comment (herbelin) [inline CHANGES]: @maximedenes: I don't know what to advise. Listing bug fixes was originally an idea from you if I remember correctly: it started for 8.5pl1 with commits from you.

For pl versions, it seems to make sense because bug fixes is what distinguishes a plX version from the previous version.

But when a list of bugs starts to contain 400 items or even a somehow arbitrary excerpt of 400 items, that we don't really have the manpower to give an appropriate description of the change for each of the bug, that no one can reasonably read all 400 items, that such a list is taking more space in the description of the changes that the structured part of the file, I start to be worried.

Maybe a discipline from each of us to associate to each fix a relevant comment in the CHANGES file at the time the fix is committed can be imagined (so that the task of reporting bug fixes is not titanic at the time of a release but incremental). Then, the change could be inserted dynamically at the right place (e.g. at the line describing "regular" subst, one may add the comment "#3070 and #5149 are also fixed"; for #3035, one could add an item in the Vernacular commands paragraph that Arguments diallow trailing arguments, mentioning #3045 in parentheses; a paragraph Efficiency could be added to host the changes reported for #3441, #3450; #3957, #4464 and many others could be simply dropped because uninformative enough to be useful, etc.).

But, yes, if we decide that this is worth the effort, it seems that, as you suggest by proposing a cep, the effort can only be a collective one.

Comment (maximedenes) [inline CHANGES]: I tend to agree with you, that the list is growing too much and not so helpful. Having structured commit messages with a "component" field would make it possible to do automatically what you suggest, and I'll try to make a proposal for the next versions.

Meanwhile, what should I do for 8.6beta1? I can still replace the list with "Many bug fixes", it could make the CHANGES file more readable. What do you think?

Comment (herbelin) [inline CHANGES]: One possibility could be to have a separate file, say BUGFIXES or so, and in which the list could go. Then, the CHANGES could have a bug section telling to look at this separate file. What do you think?

Comment (maximedenes) [inline CHANGES]: I'm starting to think that the list of bug fixes really makes little sense and is not useful, as you said. So better remove it. Plus the way I was selecting bugs is not reliable, so it could create some confusion. If nobody objects, I'll remove the list from CHANGES, tag and release.
